server/src/main/java/org/apache/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -86,7 +92,15 @@ public ResultLevelCachingQueryRunner(
     if (useResultCache || populateResultCache) {
 
       final String cacheKeyStr = StringUtils.fromUtf8(strategy.computeResultLevelCacheKey(query));
-      final byte[] cachedResultSet = fetchResultsFromResultLevelCache(cacheKeyStr);
+      DataSourceAnalysis analysis = DataSourceAnalysis.forDataSource(query.getDataSource());
+      byte[] dataSourceCacheKey = Joinables.computeDataSourceCacheKey(analysis, joinableFactory).orElse(null);
+      if (null == dataSourceCacheKey) {
+        return baseRunner.run(
+            queryPlus,
+            responseContext
+        );
+      }
+      final byte[] cachedResultSet = fetchResultsFromResultLevelCache(cacheKeyStr, dataSourceCacheKey);
```

Comment:
nit: this seems a little strange that the cache key is turned into the namespace and another key is used as the bytes, rather than making some sort of composite, but maybe it is ok?

Reply:
yeah. this is not well thought out right now. I will revisit this.

Reply:
earlier `cacheKeyStr` was being used as namespace as well as the key which was just like unnecessary duplication. Internally, the namespace and key are composed together into another bigger key. By the way, this too would cause cache bust once on an upgrade.

Reply:
This change has been removed now.